### PR TITLE
Decoder refactor

### PR DESCRIPTION
**Pull request by jindrahelcl**
Using while loop in decoder and beam search decoder. Lazy tensor annotations used somewhere.

### CODE REVIEW

**Comment by jindrahelcl** [comment]
Všechno funguje tak, jak má. 
screenshot from 2017-07-14 11 45 53

**Comment by jindrahelcl** [comment]
@jlibovicky mrkneš se na to ještě nebo ne? Klidně asi nemusíš, otestovaný to je a teď jsem se s @varisd dohodl, že udělá review on. Já se na to teď ještě podívám taky.

**Comment by jlibovicky** [comment]
Spíš se na to nepodívám :-)

File: neuralmonkey/decoders/decoder.py [hunk 1]
```diff
@@ -319,124 +374,202 @@ def get_attention_object(self, encoder, train_mode: bool):
...
+            prev_rnn_output=self.initial_state,
+            rnn_outputs=rnn_output_ta,
+            prev_logits=tf.zeros([self.batch_size, len(self.vocabulary)]),
+            logits=logit_ta,
+            prev_contexts=contexts,
+            mask=mask_ta,
+            finished=tf.zeros([self.batch_size], dtype=tf.bool),
+            attention_loop_states=attn_loop_states)
+
+    # TODO this is not exit criterions but its negation
```
Comment (varisd) [inline neuralmonkey/decoders/decoder.py]: Takze "loop_continue_criterion"? :)

File: neuralmonkey/decoding_function.py [hunk 2]
```diff
@@ -5,13 +5,29 @@
 See http://arxiv.org/abs/1606.07481
 """
 from abc import ABCMeta
+from typing import Any, Dict, List, Tuple, NamedTuple
 
 import tensorflow as tf
 from neuralmonkey.nn.projection import linear
 from neuralmonkey.nn.ortho_gru_cell import OrthoGRUCell
 
+# pylint: disable=invalid-name
```
Comment (jlibovicky) [inline neuralmonkey/decoding_function.py]: Někde tady (třeba v dokumentaci modulu) by se hodilo popsat, jaké rozhraní má ten attention. Že má ty různé historie, které vznikají různým použitím v různých dekodérech a že má ten `LoopState`, což je místo, kde se hromadí všechny side-efekty  jeho volání: hromadí se tam historie, dá se použít třeba trackování coverage.

File: neuralmonkey/decoding_function.py [hunk 3]
```diff
@@ -5,13 +5,29 @@
...
+AttentionLoopState = NamedTuple("AttentionLoopState",
+                                [("contexts", tf.TensorArray),
+                                 ("weights", tf.TensorArray)])
+# pylint: enable=invalid-name
+
+
+def empty_attention_loop_state() -> AttentionLoopState:
```
Comment (jlibovicky) [inline neuralmonkey/decoding_function.py]: Docstring.

File: neuralmonkey/decoders/decoder.py [hunk 4]
```diff
@@ -18,22 +21,39 @@
...
+from neuralmonkey.decorators import tensor
 
 RNN_CELL_TYPES = {
     "GRU": OrthoGRUCell,
     "LSTM": tf.contrib.rnn.LSTMCell
 }
 
 
-# pylint: disable=too-many-instance-attributes,too-few-public-methods
+LoopState = NamedTuple("LoopState",
```
Comment (jlibovicky) [inline neuralmonkey/decoders/decoder.py]: Možná bych v commentu napsal, co si ten `LoopState` uchovává a proč.

File: neuralmonkey/decoders/decoder.py [hunk 5]
```diff
@@ -266,42 +239,124 @@ def _create_initial_state(self) -> None:
...
+    def decoding_b(self) -> tf.Variable:
+        with tf.name_scope("output_projection"):
+            return tf.get_variable(
+                "state_to_word_b", [len(self.vocabulary)],
+                initializer=tf.constant_initializer(
+                    - math.log(len(self.vocabulary))))
+
+    @tensor
+    def train_logits(self) -> tf.Tensor:
+        # POSLEDNI TRAIN INPUT SE V DEKODOVACI FUNKCI NEPOUZIJE
```
Comment (jlibovicky) [inline neuralmonkey/decoders/decoder.py]: English

File: neuralmonkey/decoders/beam_search_decoder.py [hunk 6]
```diff
@@ -23,6 +23,16 @@
                               [("scores", tf.Tensor),
                                ("parent_ids", tf.Tensor),
                                ("token_ids", tf.Tensor)])
+
+SearchStepOutputTA = NamedTuple("SearchStepOutputTA",
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/beam_search_decoder.py]: TA historicky znamená TensorArray.. :-)

File: neuralmonkey/decoders/beam_search_decoder.py [hunk 7]
```diff
@@ -62,127 +72,190 @@ def beam_size(self):
...
+                # TODO put something else here:
+                rnn_outputs=dec_loop_state.rnn_outputs,
+                # TODO put something else here:
+                prev_logits=dec_loop_state.prev_logits,
+                # TODO put something else here:
+                logits=dec_loop_state.logits,
+                prev_contexts=bs_state.last_attns,
+                # TODO put something else here:
+                mask=dec_loop_state.mask,
+                # TODO put something else here:
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/beam_search_decoder.py]: protože když jsem to psal, tak jsem neměl rozmyšlený, co tam patří a co ne. Vůbec si timhle kódem vlastně nejsem jistej.

File: neuralmonkey/decoders/beam_search_decoder.py [hunk 8]
```diff
@@ -62,127 +72,190 @@ def beam_size(self):
...
+
+            logits = next_loop_state.prev_logits
+            state = next_loop_state.prev_rnn_state
+            attns = next_loop_state.prev_contexts
+
+            # mask the probabilities
+            # shape(logprobs) = beam x vocabulary
+            logprobs = tf.nn.log_softmax(logits)
+
+            finished_mask = tf.expand_dims(tf.to_float(bs_state.finished), 1)
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/beam_search_decoder.py]: co je vymyakovat si? :D

Comment (jlibovicky) [inline neuralmonkey/decoders/beam_search_decoder.py]: Vymaskovat :-)

File: neuralmonkey/decoders/decoder.py [hunk 9]
```diff
@@ -319,124 +374,202 @@ def get_attention_object(self, encoder, train_mode: bool):
...
+
+        embedded_input = self.embed_input_symbol(*loop_state)
+
+        return linear([embedded_input] + loop_state.prev_contexts,
+                      self.embedding_size)
+
+    def get_body(self,
+                 att_objects: List[BaseAttention],
+                 train_mode: bool,
+                 sample: bool = False) -> Callable[[LoopState], LoopState]:
```
Comment (jlibovicky) [inline neuralmonkey/decoders/decoder.py]: Výběr dalšího symbolu by se dal udělat higher-order funkcí, teď tam jsou tři možnosti: argmax, samplovat, ground truth. Taky by se to hodilo, kdybysme dělali nějaký mixture of experts.

File: neuralmonkey/decoders/beam_search_decoder.py (hunk 6 above)
Comment (jindrahelcl) [inline neuralmonkey/decoders/beam_search_decoder.py]: hotovo v docstringu modulu

File: neuralmonkey/decoding_function.py (hunk 2 above)
Comment (jindrahelcl) [inline neuralmonkey/decoding_function.py]: hotovo v docstringu modulu

File: neuralmonkey/decoding_function.py (hunk 3 above)
Comment (jindrahelcl) [inline neuralmonkey/decoding_function.py]: hotovo

File: neuralmonkey/decoders/decoder.py (hunk 9 above)
Comment (jindrahelcl) [inline neuralmonkey/decoders/decoder.py]: to bych nedával do tohohle pull requestu - stačí, že ještě neni správně ten beam search. Udělej kdyžtak issue na obecný používání higher-order funkcí. Minimálně ještě pro output_projection dekodéru, kde je sice udělanej arg pro decoder, ale nepoužívá se.

File: neuralmonkey/decoders/decoder.py [hunk 10]
```diff
@@ -266,42 +239,124 @@ def _create_initial_state(self) -> None:
...
+        return tf.contrib.seq2seq.sequence_loss(
+            tf.transpose(self.train_logits, perm=[1, 0, 2]),
+            train_targets,
+            tf.transpose(self.train_padding),
+            average_across_batch=False)
+
+    @tensor
+    def train_loss(self) -> tf.Tensor:
+        return tf.reduce_mean(self.train_xents)
+
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/decoder.py]: `train_xents` se počítaj na transponovaných `train_inputs`, protože `train_inputs` maj v první dimenzi čas a v druhý batch. `tf.contrib.seq2seq.sequence_loss` žere jako targety tensor, co má v první dimenzi batch a ve druhý čas. Tahle funkce je schopná vracet loss pro každý slovo. My chcem, aby `train_xents` byly lossy pro každou větu. Třeba kvůli nějakejm reinforce hárdkórech. A celej `train_loss` už je pak jen  skalár. Ty `train_xents` jsou umělej mezikrok. Kdyby tam nebylo to `average_across_batch=False`, tak by to bylo to samý jako tadyten loss.

File: neuralmonkey/decoders/beam_search_decoder.py (hunk 7 above)
Comment (jindrahelcl) [inline neuralmonkey/decoders/beam_search_decoder.py]: Ne - jsem si jistej, že je špatně. :-) Bere věci tak, jak je vyplivnul decoder a nebere v potaz, že se s tim nějak hejbe.
